File: python/leetcode/Two Sum.py
class Solution:
    def memoziation(self, nums: List[int], target: int) -> List[int]:
        mem = dict()
        for i, v in enumerate(nums):
            j = mem.get(v, -1)
            if j >= 0:
                return [i, j]
            else:
                mem[target - v] = i

    def twoSum(self, nums: List[int], target: int) -> List[int]:
        return self.memoziation(nums, target)

File: java-kotlin/src/others/TwoSum.java
package others;

import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> mem = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];

            if (mem.containsKey(diff)) {
                return new int[] { mem.get(diff), i };
            } else {
                mem.put(nums[i], i);
            }
        }

        throw new IllegalArgumentException("No solution");
    }
}